NOTES-daneroo.md

```diff
@@ -0,0 +1,104 @@
+# Notes for PR(daneroo) feature/listing branch
+
+I don't expect this to be merged as-is. I mostly wanted to share my experiments and get feedback.
+
+If any of it is useful, I can craft proper commits with smaller features.
+
+*I am a first-time contributor, and not very experienced with Go, so would appreciate any feedback both on content and process.*
+
+## Overview
+
+When I first started to use the project, I experienced instability and would need to restart the download process a large number of times to get to the end. The process would timeout after 1000-4000 images. To test, I use two accounts one with ~700 images, and one with 31k images.
+
+The timeouts were mostly coming from the `navLeft()` iterator, so I disabled actual downloading and focused on traversal. (`-list` option). I also added an option (`-all`) to bypass the incremental behavior of `$dldir/.lastDone`. I then added `-vt (VerboseTiming`, to isolate timing metrics reporting specifically.
+
+The last main issue was the degradation of performance over time. Probably due to resource leaks in the webapp, especially when iterating at high speed. I found that the simplest solution to this was simply to force a reload every 1000 iterations. This typically takes about `1s` and avoids the `10X` slowdown I was observing.
+
+The following are details of each of the major changes.
+
+## `navLeft()`
+
+- By adding the `-list` option, its functionality is isolated.
+- Remove `WaitReady("body",..)`, as it has no effect (The document stays loaded when we navigate from image to image), and replace it with a loop (until `location != prevLocation`).
+- When only listing, this is the critical part of the loop, so I spent some time experimenting with the timing of the chromedp interactions to optimize total throughput. What seemed to work best was an exponential backoff, so that we can benefit from the *most likely* fast navigation, but don't overtax the browser when we experience the occasional longer navigation delay (which can go up to multiple seconds).
+
+- We also pass in `prevLoaction` from `navN()`, to support the new `navLeft()` loop, to optimize the throughput.
+
+## `navN()`
+
+- Addition of timing code
+- Support listing only (`-list`)
+- Reload page every (`batchSize=1000`) items.
+
+## Termination Criteria
+
+We explicitly fetch the lastPhoto, on the main album page (immediately after authentication), and rely on that, instead of `navLeft()` failing to navigate (`location == prevLocation`). The lastPhoto is determined with a CSS selector evaluation (`a[href^="./photo/"]`), which is fast and should be stable over time, as it uses the `<a href=.. >` detail navigation mechanism of the main album page. This opens another possibility for iteration on the album page itself. (see `listFromAlbum` below)
```

Comment:
Regarding that change, can it work for the "main page" whole gallery (for lack of a better term) as well, and not just for a specific album? If yes, I think we want it, and that you could make a PR for that if you want.
Even better, you may have noticed the worst part of this program, which is that to jump at the very beginning of the feed (oldest photo), we're continuously scrolling down and doing screenshots of the screen, until the screenshot does not change anymore. I know, disgusting. So if you know how to jump for sure to the oldest element of the feed with a CSS selector or other such trick, that would be really great. Are you interested in doing any of that?

Reply:
Yes, the way the web app is structured, I consider the *Main Page* to be a special album page. I use the term *Album Page* in contrast with the *Detail Page* which shows a single photo. So I could make a separate PR, which would identify the `lastPhoto` (newest) before we `NavToEnd`, and then replace the termination criteria with reaching that photo.

With respect to the scrolling, I don't really have another option to `kb.PageDown + kb.End`. Ther content of the Album/Main Page can be seen as a virtual paging window over the collection, essentially the DOM only contains a few more images (before and after) than those that are currently visible.

However, I did try an alternative to the `CaptureScreenshot`, which finds the last photo in the DOM which is faster than the screenshot, but doesn't really remove the icky *Am I at the end* loop. Let's put off the `NavToEnd` stuff for later.

Reply:
> Yes, the way the web app is structured, I consider the _Main Page_ to be a special album page. I use the term _Album Page_ in contrast with the _Detail Page_ which shows a single photo. So I could make a separate PR, which would identify the `lastPhoto` (newest) before we `NavToEnd`, and then replace the termination criteria with reaching that photo.

Excellent.

> With respect to the scrolling, I don't really have another option to `kb.PageDown + kb.End`. 

That's fine. I'm actually fairly satisfied with that part. it's only the stopping criterion I'm sad about.

> Ther content of the Album/Main Page can be seen as a virtual paging window over the collection, essentially the DOM only contains a few more images (before and after) than those that are currently visible.
> 
> However, I did try an alternative to the `CaptureScreenshot`, which finds the last photo in the DOM which is faster than the screenshot, but doesn't really remove the icky _Am I at the end_ loop. Let's put off the `NavToEnd` stuff for later.

as you wish, but I think that might be good enough as a first PR though.